src/gig_argument.c

```diff
@@ -678,64 +678,184 @@ scm_to_c_ptr_array(S2C_ARG_DECL)
+                _meta.transfer = GI_TRANSFER_EVERYTHING;
+
+                gig_argument_scm_to_c(subr, argpos, &_meta, keyval[j], must_free, &_keyval[j],
+                                      &_size);
+
+                // GHashTables apparently expect that negative
+                // integers have intptr_t sign extension.
+                if ((j == 0 && key_type == GIG_HASH_INT)
+                    || (j == 1 && val_type == GIG_HASH_INT)) {
+                    gpointer p;
+                    int item_siz;
+                    item_siz = meta->params[j].item_size;
+                    if (item_siz == 1)
+                        p = GINT_TO_POINTER(_keyval[j].v_int8);
+                    else if (item_siz == 2)
+                        p = GINT_TO_POINTER(_keyval[j].v_int16);
+                    else if (item_siz == 4)
+                        p = GINT_TO_POINTER(_keyval[j].v_int32);
+                    else
+                        g_assert_not_reached();
+                    _keyval[j].v_pointer = p;
+                }
+                else if ((j == 0 && key_type == GIG_HASH_INT64) ||
+                         (j == 1 && val_type == GIG_HASH_INT64)) {
+                    // GHashTables expect gint64 to be passed by
+                    // reference, even if they fit in a pointer.
+                    if (_meta.gtype == G_TYPE_INT || _meta.gtype == G_TYPE_INT64) {
+                        gint64 *p = g_malloc(sizeof(gint64));
+                        *p = _keyval[j].v_int64;
+                        _keyval[j].v_pointer = p;
+                    }
+                    else if (_meta.gtype == G_TYPE_UINT || _meta.gtype == G_TYPE_UINT64) {
+                        guint64 *p = g_malloc(sizeof(guint64));
+                        *p = _keyval[j].v_uint64;
+                        _keyval[j].v_pointer = p;
+                    }
+                }
+                else if ((j == 0 && key_type == GIG_HASH_REAL) ||
+                         (j == 1 && val_type == GIG_HASH_REAL)) {
+                    // GHashTables expect double and float to be
+                    // passed by reference, even if they fit in a
+                    // pointer.
+                    if (_meta.gtype == G_TYPE_DOUBLE) {
+                        gdouble *p = g_malloc(sizeof(double));
+                        *p = _keyval[j].v_double;
+                        _keyval[j].v_pointer = p;
+                    }
+                    else if (_meta.gtype == G_TYPE_FLOAT) {
+                        // There is no known GObject library uses
+                        // hashtables of floats, so this is for values
+                        // only.
+                        gfloat *p = g_malloc(sizeof(double));
+                        *p = _keyval[j].v_float;
+                        _keyval[j].v_pointer = p;
+                    }
+                    else
+                        g_assert_not_reached();
+                }
```

Comment:
I'd abstract this part into a function, so that you can call it once with `key_type` and once with `value_type`.